src/jit/decompose.cpp

```diff
@@ -0,0 +1,1189 @@
+        DecomposeCnsLng(ppTree, data);
+        break;
+
+    case GT_CALL:
+        DecomposeCall(ppTree, data);
+        break;
+
+    case GT_RETURN:
+        assert(tree->gtOp.gtOp1->OperGet() == GT_LONG);
+        break;
+
+    case GT_STOREIND:
+        DecomposeStoreInd(ppTree, data);
+        break;
+
+    case GT_STORE_LCL_FLD:
+        assert(tree->gtOp.gtOp1->OperGet() == GT_LONG);
+        NYI("st.lclFld of of TYP_LONG");
+        break;
+
+    case GT_IND:
+        NYI("GT_IND of TYP_LONG");
+        break;
+
+    case GT_NOT:
+        DecomposeNot(ppTree, data);
+        break;
+
+    case GT_NEG:
+        NYI("GT_NEG of TYP_LONG");
+        break;
+
+    // Binary operators. Those that require different computation for upper and lower half are
+    // handled by the use of GetHiOper().
+    case GT_ADD:
+    case GT_SUB:
+    case GT_OR:
+    case GT_XOR:
+    case GT_AND:
+        DecomposeArith(ppTree, data);
+        break;
+
+    case GT_MUL:
+        NYI("Arithmetic binary operators on TYP_LONG - GT_MUL");
+        break;
+
+    case GT_DIV:
+        NYI("Arithmetic binary operators on TYP_LONG - GT_DIV");
+        break;
+
+    case GT_MOD:
+        NYI("Arithmetic binary operators on TYP_LONG - GT_MOD");
+        break;
+
+    case GT_UDIV:
+        NYI("Arithmetic binary operators on TYP_LONG - GT_UDIV");
+        break;
+
+    case GT_UMOD:
+        NYI("Arithmetic binary operators on TYP_LONG - GT_UMOD");
+        break;
+
+    case GT_LSH:
+    case GT_RSH:
+    case GT_RSZ:
+        NYI("Arithmetic binary operators on TYP_LONG - SHIFT");
+        break;
+
+    case GT_ROL:
+    case GT_ROR:
+        NYI("Arithmetic binary operators on TYP_LONG - ROTATE");
+        break;
+
+    case GT_MULHI:
+        NYI("Arithmetic binary operators on TYP_LONG - MULHI");
+        break;
+
+    case GT_LOCKADD:
+    case GT_XADD:
+    case GT_XCHG:
+    case GT_CMPXCHG:
+        NYI("Interlocked operations on TYP_LONG");
+        break;
+
+    default:
+        {
+            JITDUMP("Illegal TYP_LONG node %s in Decomposition.", GenTree::NodeName(tree->OperGet()));
+            noway_assert(!"Illegal TYP_LONG node in Decomposition.");
+            break;
+        }
+    }
+
+#ifdef DEBUG
+    if (comp->verbose)
+    {
+        printf("  AFTER:\n");
+        comp->gtDispTree(*ppTree);
+    }
+#endif
+}
+ 
+
+//------------------------------------------------------------------------
+// FinalizeDecomposition: A helper function to finalize LONG decomposition by
+// taking the resulting two halves of the decomposition, and tie them together
+// with a new GT_LONG node that will replace the original node.
+//
+// Arguments:
+//    ppTree - the original tree node
+//    data - tree walk context
+//    loResult - the decomposed low part
+//    hiResult - the decomposed high part
+//
+// Return Value:
+//    None.
+//
+void Decompose::FinalizeDecomposition(GenTree** ppTree, Compiler::fgWalkData* data, GenTree* loResult, GenTree* hiResult)
```

Comment:
Side note: this might turn out to be not very useful as it's not very flexible in its handling of `hiResult`. We'll see.